tests/scripts/collect-logs.sh

```diff
@@ -22,13 +22,8 @@ if [[ "$OPERATOR_NAMESPACE" != "$CLUSTER_NAMESPACE" ]]; then
   NAMESPACES+=("$OPERATOR_NAMESPACE")
 fi
 
-if [ "$1" == "multi-cluster-mirroring" ]; then
-  NAMESPACES+=("rook-ceph-secondary")
-fi
-
-# Add kube-system namespace for multus test only as we need to debug network in multus test
-if [ "$1" == "canary-multus" ]; then
-  NAMESPACES+=("$KUBE_SYSTEM_NAMESPACE")
+if [[ -n "${ADDITIONAL_NAMESPACE}" ]]; then
+  NAMESPACES+=("${ADDITIONAL_NAMESPACE}")
 fi
```

Comment:
@subhamkrai since you work on and review a lot of the CI code, I'd like to ask for your help to make sure that the code coming into the CI is more robust and resilient to human errors. I saw something in the scripts that isn't a huge deal right now, but it is a trend that can lead to problems in the long-term if we don't keep code cleanliness in mind, and some help keeping CI cleanliness in mind would be much appreciated.

The two short sections of code tagged here were "correct" when created, but I don't think they were "robust." I'll do my best to explain what I'm seeing, although it is a complex and nuanced issue.

In this scenario, code at different abstraction levels was silently "coupled" together. When I see that a higher-level property (in this case, the name of the test suite, `canary-multus` or `multi-cluster-mirroring`) is used as a conditional statement in a lower-level function AND the property isn't passed explicitly to the low-level function, that is a "code smell". It isn't always wrong, but it is a hint that the overall system (CI in this case) might be fragile, and it should prompt us to consider whether another route can be taken.

One of the ways that this code had already failed between when it was added and my PR here is the `canary-multus` conditional: there is no `canary-multus` test today. I think I was probably the one to rename `canary-multus` to `multus-cluster-network`, but because I wasn't aware of the code in this script, I caused CI to silently stop collecting pod logs from the `kube-system` namespace. And because others weren't aware of the requirement from the script, they missed the bug as well. 

To help create more a more robust CI, what I have proposed here is to make the additional namespace collection explicit by passing a more generic ADDITIONAL_NAMESPACE env var parameter from GH actions (through the collect-logs layer) to this script. This propagation allows the CI test that does the calling to be the one to codify what extra namespace needs to be collected. When the calling CI is modified or renamed, there is less risk that the lower-level functionality will break silently because it is no longer silently coupled -- it's explicitly coupled.

As the code is written in this PR, if I change the caller to remove the extra namespace on accident, it would be more likely that someone would catch the error because it would show up in the diff.

The risk of silently coupled code is that when we need to make small high-level shifts to the code, we end up with low-level changes also needing to be made. If different contributors all create code like this that we don't clean up, then code at various hierarchical levels ends up all coupled to code at different levels in a complex and hard to understand flow. This is "spaghetti code" where it becomes harder and harder to change one thing because everything requires changes to something else.

It's also worth noting that one instance of this kind of error usually doesn't totally break CI. The impact of this bug was that if the multus test were failing, we would need to fix the CI to get the correct logs again. That isn't a huge deal, but would make the process of debugging the issue longer and more frustrating.

Finally, there is the nuanced part of this. The hardest part of coding is knowing when it's safe to make an exception to a rule. There are times where we might choose to ignore silent coupling because passing parameters around is more work than it seems likely to save. For a concrete example, this could be a parameter like the system architecture. If a CI test runs once for x86 and once for ARM, we might not pass the `arch` parameter down through the layers because it is so easy to query the system architecture at any software layer. There are still ways in which our "assumptions" might break over time. For example, if the ARM run is modified to be done using a VM on an x86 host, this might break the system silently if parts of the scripting return x86 when they should return ARM. However, we might decide to take that risk and do less work now because (1) passing a basic param around can add confusion, (2) changing the arch to VMs isn't very likely, and (3) it is probably fairly easy to change to a VM later and still not need to pass the arch around from the top level.

Reply:
I totally agree with you regarding the small small issues we have in the github action and related bash scripts. In general, bash script are harder to write clean code and more often we concentrate on the CI files when something needs to be fixed or something needs to be added. I think we mostly skip focusing on we can help have less duplicate and more reliable code. I can this behaviour of code to myself as mostly I tried to fixed the CI issue or add instead of optimising. And also with time CI files have grown much and as you mentioned it get's really difficult to reduce the complexity. It's good time to review the CI files as we review go code to  have less duplicates and more readable files. Thanks for bringing this.